FEXCore/Source/Utils/FutexSpinWait.h

```diff
@@ -0,0 +1,282 @@
+   * by putting the CPU in to a lower-power state using WFE.
+   * On platforms tested, WFE will put the CPU in to a lower power state for upwards of 52ns per WFE. Which isn't a significant amount of time
+   * but should still have power savings. Ideally WFE would be able to keep the CPU in a lower power state for longer. This also has the added benefit
+   * that atomics aren't abusing the caches when spinning on a cacheline, which has knock-on powersaving benefits.
+   *
+   * FEAT_WFxT adds a new instruction with a timeout, but since the spurious wake-up is so aggressive it isn't worth using.
+   *
+   * It should be noted that this implementation has a few dozen cycles of start-up time. Which means the overhead for invoking this implementation is
+   * slightly higher than a true spin-loop. The hot loop body itself is only three instructions so it is quite efficient.
+   *
+   * On non-ARM platforms it is truly a spin-loop. Which is okay for debugging only.
+   */
+  class FutexSpinWait final {
+    public:
+#ifdef _M_ARM_64
+      template<typename T, typename TT>
+      static void Wait(T *Futex, TT ExpectedValue) {
+        std::atomic<T> *AtomicFutex = reinterpret_cast<std::atomic<T>*>(Futex);
+        T Tmp{};
+        T Result = AtomicFutex->load();
+
+        // Early exit if possible.
+        if (Result == ExpectedValue) return;
+
+        if (!StaticDataInitialized.load(std::memory_order_relaxed)) [[unlikely]] Init();
+
+        do {
+          if constexpr (sizeof(T) == 1) {
+            __asm volatile(SPINLOOP_8BIT
+              : [Result] "=r" (Result)
+              , [Tmp] "=r" (Tmp)
+              : [Futex] "r" (Futex)
+              , [ExpectedValue] "r" (ExpectedValue)
+              : "memory");
+          }
+          else if constexpr (sizeof(T) == 2) {
+            __asm volatile(SPINLOOP_16BIT
+              : [Result] "=r" (Result)
+              , [Tmp] "=r" (Tmp)
+              : [Futex] "r" (Futex)
+              , [ExpectedValue] "r" (ExpectedValue)
+              : "memory");
+          }
+          else if constexpr (sizeof(T) == 4) {
+            __asm volatile(SPINLOOP_32BIT
+              : [Result] "=r" (Result)
+              , [Tmp] "=r" (Tmp)
+              : [Futex] "r" (Futex)
+              , [ExpectedValue] "r" (ExpectedValue)
+              : "memory");
+          }
+          else if constexpr (sizeof(T) == 8) {
+            __asm volatile(SPINLOOP_64BIT
+              : [Result] "=r" (Result)
+              , [Tmp] "=r" (Tmp)
+              : [Futex] "r" (Futex)
+              , [ExpectedValue] "r" (ExpectedValue)
+              : "memory");
+          }
+          else {
+            static_assert(!std::is_same_v<T, T>, "Invalid");
+          }
+        } while (Result != ExpectedValue);
+      }
+
+      template<typename T, typename TT>
+      static bool Wait(T *Futex, TT ExpectedValue, std::chrono::nanoseconds const &Timeout) {
```

Comment:
I'm guessing these are implemented directly to allow inlining, however it's very difficult to actually see what interface is being exposed that way. It would help to forward-declare all of these functions in a single block hence.